modules/context/context.go

```diff
@@ -387,6 +388,15 @@ func (ctx *Context) SetServeHeaders(opts *ServeHeaderOptions) {
 		backslashEscapedName := strings.ReplaceAll(strings.ReplaceAll(opts.Filename, `\`, `\\`), `"`, `\"`) // \ -> \\, " -> \"
 		header.Set("Content-Disposition", fmt.Sprintf(`%s; filename="%s"; filename*=UTF-8''%s`, disposition, backslashEscapedName, url.PathEscape(opts.Filename)))
 		header.Set("Access-Control-Expose-Headers", "Content-Disposition")
+
+		if len(setting.Repository.AccessControlAllowOrigin) > 0 {
+			origin := strings.ToLower(ctx.Req.Header.Get("Origin"))
+			matcher := regexp.MustCompile(`http(s?)://`)
+			origin = matcher.ReplaceAllString(origin, "")
+			if origin == setting.Repository.AccessControlAllowOrigin {
+				header.Set("Access-Control-Allow-Origin", setting.Repository.AccessControlAllowOrigin)
+			}
+		}
```

Comment:
TBH, it doesn't look good to me.

Although I think we can respect the Repository.AccessControlAllowOrigin config, but this logic seems strange, it's different from how the AccessControlAllowOrigin is used in `httpBase`.

Reply:
> Although I think we can respect the Repository.AccessControlAllowOrigin config, but this logic seems strange, it's different from how the AccessControlAllowOrigin is used in `httpBase`.

You mean removing of a protocol prefix, don't You? I'm agree that logics should be the same. Do You think that this removing is excess?

Reply:
> I believe every single line should have its clear purpose and meaning. If httpBase is correct, then new code should follow it (extract the common part to a new function, to avoid duplicate code). If your new code is correct, then is old httpBase incorrect and should be fixed? If httpBase and your code are both correct, then what's the difference, why they should be written differently (which causes inconsistent behavior for the single config option)?

You're quite right! But I see the following facts:
1. `http.go` implements Git Smart HTTP protocol. This module doesn't take part in responsing to usual requests from web browser or other simple clients.
2. `httpBase` sends `Access-Control-Allow-Origin` to all clients. It's wrong. Someone should correct it by another PR. I'm better just to add TODO remark in `http.go`, I think.
3. `http.go` and `context.go` really have some similar code. This [fragment](https://github.com/go-gitea/gitea/blob/main/routers/web/repo/http.go#L48-L70) could be located in a separate function. But I think it is not useful right now. It should be a part of further refactoring. But I can add one more TODO remark about this idea.

What do You think?

Reply:
At least, according to MDN: `Access-Control-Allow-Origin: <origin>` , https://developer.mozilla.org/en-US/docs/Glossary/Origin , the code `origin = matcher.ReplaceAllString(origin, "")` doesn't look right.

Reply:
To fix the `httpBase` problem, I proposed #24303 , using a standard CORS package could be better.

Reply:
OK, I'll wait for https://github.com/go-gitea/gitea/pull/24303

Reply:
ps: usually the PR doesn't need to sync with main frequently (unless it is out-dated for long time, or there are conflicts). Just leave it there, as long as the PR is ready and approved, maintainers could help to sync & update & merge.

Reply:
#24303 has been merged